### ISSUE TITLE
Image Block: When adding an existing image within a post, explore "attaching" the image to the post if it isn't already attached

### ISSUE THREAD

**Issue by andrewserong** [WordPress/gutenberg#66663]
## What problem does this address?

Follows #12753

This issue was raised as part of a much earlier issue (#12753) in this comment: https://github.com/WordPress/gutenberg/issues/12753#issuecomment-556149071

> Worth noting here that some work would need to be done to ensure Gutenberg added images are properly seen as attached.
> Right now, you can have an image in a Gutenberg post or page but the status in WP Admin > Media will still show unattached.

In Gutenberg, if we upload an image to a particular post, then the uploaded image will be "attached" to that post. I.e. if you go to the media library, you can see that it's attached to the post:

![Image](https://github.com/user-attachments/assets/c7945029-ce2d-465a-b3d1-533667ccbb16)

However, if we simply select an image from the media library and add it to a post or page, it does not "attach" it to the post.

To verify, add some images to a site via the media library. Then, go to create a post or page. Add the image to the post or page and publish. Then, go to the media library, and the image will still say `Unattached` under the `Uploaded to` column:

![Image](https://github.com/user-attachments/assets/c874b38a-50eb-426f-a04a-24276a1824c3)

## What is your proposed solution?

When a user adds an image from the media library to a post or page, if the image is not already attached to any post (is unattached), then attach the image to the post.

That's the proposal here! Does this sound good, or are there any potential gotchas, concerns, issues that could arise by introducing this?

**Comment by Marc-pi** [comment]
yep, i stumbled upon this case this week on a customer website, deleting some media flagged as unattached, but were not ;-)

**Comment by Mamaduka** [comment]
Somewhat related to the Media Library behavior - #17658.

**Comment by ellatrix** [comment]
Imo the whole concept of media attachment should be rethought. Images can be attached to multiple posts, it would be cool to surface all of them.

**Comment by Vrishabhsk** [comment]
Hi @andrewserong @ellatrix @Mamaduka 
- I have raised a PR for the feature's first iteration : https://github.com/WordPress/gutenberg/pull/70313

Let me know the further steps on this and if the multiple posts feature also needs to be explored. Thanks

**Comment by ramonjd** [comment]
I've been looking at this also today, and have been trying to work out a way to assign the post parent.

I'm not sure what the right way to do this would be.

Firstly, I'm assuming the following:

- Assigning a parent post id where none exists on attachment selection
- It should work for all attachments, not just images
- It should work for gallery images

Here's what works locally so far. In the [onSelect](https://github.com/WordPress/gutenberg/blob/trunk/packages/media-utils/src/components/media-upload/index.js#L432) callback of `<MediaUpload />` make a call to an async utility function:

```js
onSelect() {
		// ... code above
		const { wp } = window;
		if ( attachment?.length && wp.media.view.settings.post.id ) {
			assignParentPost( attachment, wp.media.view.settings.post.id );
		}
}
```

`assignParentPost` looks something like this:

```ts
/**
 * Internal dependencies
 */
import type { RestAttachment } from './types';
import { transformAttachment } from './transform-attachment';

export async function assignParentPost(
	attachments: RestAttachment[],
	parentPostId: RestAttachment[ 'id' ],
	signal?: AbortSignal
) {
	const promises = attachments.map( async ( attachment ) => {
		if ( ! attachment.id || !! attachment.uploadedTo ) {
			return null;
		}
		const response = await apiFetch< RestAttachment >( {
			path: `/wp/v2/media/${ attachment.id }`,
			body: JSON.stringify( {
				post: parentPostId, // This sets the post_parent field
				alt_text: 'nobby',
			} ),
			method: 'PUT',
			signal,
		} );
		return transformAttachment( response );
	} );

	return Promise.all( promises );
}
```

It's crude but it works.

There are, however, a couple of questions/challenges I've noted:

- the assignment takes place before the user saves the post, which means the user could select an image, then decide to remove it.
- it relies on Core's backbone model. I only mention it because I don't trust what I don't understand 😆 
- it's not hooked into the core data store -  updating a record should update the store.

We could think about this another way and ask: 

- are there any attachments in the post content, and, if so 
- do those attachments have parents, and if not
- attach them after/on saving the post

That throws up a whole new set of tasks, namely where to do all this - in the attachment blocks? In the core-data store? I'm really not sure, but I'm thinking it should be central rather than distributed. 

By that I mean, any newly-inserted attachments should be updated when the post is saved. Uploaded image are already assigned the `post` value, so we really only care about selected images. Following that logic, we'd need to flag them as having been selected, and then check if they're still in the post when the image is saved.

Maybe `media-utils` could track the selected attachments in some internal store that core-data could access when the post is saved. 🤷🏻 I'm just riffing here.

**Comment by andrewserong** [comment]
>  I'm just riffing here.

I'm enjoying the riffing, thanks for laying out your thoughts!

I like the idea of centralizing where the updates occur. One challenge with leaving it until save/publish time is that on a lot of blogs there might be a very large number of images (I just checked one of my own blog posts and found it had 63 image blocks in it). So depending on the size of the post, if we wait until save/publish, it could be a large volume of API requests and slow down publishing 🤔

I'm also not sure what the right solution would be there, but just thought I'd add the thought!

**Comment by ramonjd** [comment]
>  So depending on the size of the post, if we wait until save/publish, it could be a large volume of API requests and slow down publishing

Good point. Maybe it could be a new endpoint in the attachments controller. So the idea would be to store a set of ids of inserted images in the client, and then send them off for processing in the backend. 

I can't tell, but I get the feeling it'd be easier to reason about if the media modal were better integrated with WordPress components/core data store. That way we'd be able to interact with the entities store and it'd all be updated on the fly. 

I think.

**Comment by ramonjd** [comment]
> Imo the whole concept of media attachment should be rethought. Images can be attached to multiple posts, it would be cool to surface all of them.

Meanwhile, 15+ years ago... 😄 

- [Associate Attachments with Multiple Parents](https://core.trac.wordpress.org/ticket/15471)
- [Time for a wp_post_relationships table?](https://core.trac.wordpress.org/ticket/14513)
- [Allow many-to-many relationship between posts and attachments](https://core.trac.wordpress.org/ticket/10657)

**Comment by talldan** [comment]
Just dropping some thoughts. I personally think the code that maintains the relationships between attachment and post should be server-side and not on the client. In WordPress there are myriad ways to add or remove media and posts (e.g. wp-cli, non-core editors, plugins, tools), so keeping the code as close to the source of truth as possible means it's more likely to work for these wider uses.

I do agree though with the earlier observation that media attachment doesn't really serve it's purpose. It seems like should be considered part of this issue [Tracking and providing insights about consumption of global elements](https://github.com/WordPress/gutenberg/issues/60205#top) for the longer term solution.

**Comment by andrewserong** [comment]
Certainly worth exploring a server-side solution, that idea's got a lot of merit! I'm not sure what's happening on the backend, but I notice in the media library that if I have media that has a parent, and then delete the post it's attached to, when I reload the media library, it shows unattached again. So it looks like that side of things might already be (at least partially) handled:

Before deleting "a test page":

<img width="1105" height="141" alt="Image" src="https://github.com/user-attachments/assets/20c6a860-d1e6-450b-af7b-95a0b2706aff" />

After deleting "a test page":

<img width="1094" height="134" alt="Image" src="https://github.com/user-attachments/assets/c2ea27b6-2c04-4ede-99ad-223611a3bc31" />

Edit: from the looks of it, I think the UI side in core / the classic media library handles it by checking for the existence of the parent before displaying the attached status in the UI: https://github.com/WordPress/wordpress-develop/blob/b4418472120158fd9a802c7343f7dbffa31c9862/src/wp-includes/media.php#L4545-L4551 (so if the post has been deleted, it'll display as unattached)

**Comment by andrewserong** [comment]
Also, I just looked up the media library code in use in the Classic Editor, which appears to do an AJAX call to update the post parent when a user selects an image from the media library while editing a post: https://github.com/WordPress/wordpress-develop/blob/9ed92faaca92f66ec9e2979c9d896ddab3e4e993/src/js/_enqueues/wp/media/editor.js#L983-L988

So, it seems the current state of things (as far as I can tell) in core is that:

* In the Classic Editor, a post parent was assigned when a user selects an image from the media library, but before the post itself is published.
* The media library ignores any "Uploaded to" / post parent value where the post doesn't exist.

This means that media in the database has stale post parent values, and things appear to be handled by the UI

---

(I'm not saying we should keep the status quo or implement the same in the block editor, but I was just curious to see how things were set up prior to Gutenberg).

**Comment by andrewserong** [comment]
Update: I've started an exploratory PR to try out the idea of attaching media client-side on selection (attempting to try out parity between the block editor and what the classic editor was doing previously). It's likely one of the simplest ways to try this out, but also has its drawbacks (the action is performed silently, there's no undo, and the attaching behaviour occurs before publish/save):

* https://github.com/WordPress/gutenberg/pull/81804

I'll also try out an alternative to explore the server-side idea: I suspect there'll be drawbacks of any approach, though. I.e. if we wait until publish, then there won't be real-time feedback in the editor when an image is attached, which is now more noticeable with the Dynamic Gallery block. For example, if I have a Dynamic Gallery block on the page and then add an image to that page, I'd expect the Dynamic Gallery block to update immediately, rather than having to wait until I've saved or published the post 🤔

**Comment by andrewserong** [comment]
Alrighty, I've opened up a more complex alternative to #81804 over in:

* https://github.com/WordPress/gutenberg/pull/81856

This one is my attempt to hone in on a UX that gives the user the most control (at the expense of increased code complexity). In this one we:

* Check the post in the pre-publish panel to see which images in the post need to be attached
* Show the user a panel where they can de-select images they don't want to attach (collapsed by default)
* After the post is published, use edited entity record updates to flag that added unattached images will be attached at save

This defers things to publish time while still giving user control, but it is a fair bit more complex!

The more I dig into this, though, the less inclined I am to explore a purely server-side approach as I think this feature is really about how the block editor manages media, and I think it's more to do with images added as part of Image and Gallery blocks where we'd expect attached images to show up, but not images like background images, cover block images and so on.
